File: my-folder/0020-valid-parentheses/solution.java
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) return false;
        Stack<Character> st = new Stack<>();
        char top;
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') st.push(c);
            else {
                if (st.isEmpty()) return false;
                else {
                    top = st.pop();
                    switch (c) {
                        case ')':
                            if (top != '(') return false;
                            break;
                        case ']':
                            if (top != '[') return false;
                            break;
                        case '}':
                            if (top != '{') return false;
                            break;
                    }
                }
            }
        }
        if (!st.isEmpty()) return false;
        return true;
    }
}

File: my-folder/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for char in s:
            if char == '(' or char == '[' or char == '{':
                stack.append(char)
            else:
                if not stack: return False
                top = stack[-1]
                if (char == ')' and top == '(') or (char == ']' and top == '[') or (char == '}' and top == '{'):
                    stack.pop()
                else:
                    stack.append(char)

        return not stack

